ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,933 @@
+/// a `NIDZKP` which is computed over multiple recursions , is a vector of `g` polynomials
+/// the length is the amount of proof recursions
+/// `g` is secret shared using PRSS: left share is stored in struct, right share is generated via PRSS
+/// the length of the left share, `g_left`, determines the degree of the polynomial
+/// for a low communication rate, i.e. `log (amount of multiplications)`,
+/// the length needs to be constant in the amount of multiplications
+/// length might be different across recursions (therefore used `Vec` instead of `GenericArray`)
+/// the "left proof" also contains non-zero masks, `mask_p` and `mask_q`.
+/// For the "right proof" both masks are zero.
+///
+/// Proofs that are sent to other parties have mask that are always zero and therefore don't need to be sent.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct NIDZKP<F>
+where
+    F: Field,
+{
+    pub(crate) proofs: Vec<Vec<F>>,
+    pub(crate) mask_p: F,
+    pub(crate) mask_q: F,
+}
+
+/// generates recursive proof for `sum_i u_i*v_i = out` over field `F`
+///
+/// `r_f`, i.e. recursion factor, determines how much the inputs are compressed during each recursion
+/// it is also the degree of polynomials p and q
+///
+/// the output on the right needs to be sent to the party on the left
+/// it is therefore the "right share" of the party on the left
+///
+/// the left part of the output is the "left share" of the proof generated by the party on the right
+/// this part is generated using PRSS and does not need to be sent,
+/// it remains with the party and is the input `proof_left` to the proof verification
+/// the outputs includes the two masks for the proofs, i.e. `mask_p`, `mask_q`
+///
+/// This function uses `ctx` for generating random values using PRSS
+/// and sending the right part of the proof to the left
+///
+/// # Errors
+/// propagates errors from send and receive
+///
+/// # Panics
+/// Does not panic
+pub async fn generate_proof<C, F>(
+    ctx: C,
+    u: &mut Vec<F>,
+    v: &mut Vec<F>,
+    r_f: usize,
+) -> Result<(NIDZKP<F>, NIDZKP<F>), Error>
+where
+    C: Context,
+    F: Field + Invert,
+{
+    // check length
+    debug_assert_eq!(u.len(), v.len());
+    // check that field is large enough to hold evaluation points
+    debug_assert!(F::BITS > usize::BITS - r_f.leading_zeros() + 1);
+    // record, used as counter for PRSS
+    let mut record = 0usize;
+
+    // proof length
+    #[allow(
+        clippy::cast_possible_truncation,
+        clippy::cast_sign_loss,
+        clippy::cast_precision_loss
+    )]
+    let proof_length = (u.len() as f64).log(r_f as f64).ceil() as usize + 1usize;
+    // get masks from PRSS
+    let mask_p = ctx.prss().generate_fields(RecordId::from(record)); //(F::ONE, F::ONE);
+    record += 1;
+    let mask_q = ctx.prss().generate_fields(RecordId::from(record));
+    // proof from the prover on the left, i.e. generated via `PRSS.left` rather than receiving it from the left
```

Comment:
We are inside the function `generate_proof` so it's kind of strange to talk about "receiving it" from anyone. This code is run by the prover, not received from anyone else.